chatmeleon-server/src/controllers/message-controller.ts

```diff
@@ -1,114 +1,252 @@
- * Whether the conversation exists and the user is a member of it.
- */
-const hasConversation = async (
-  userId: string,
-  conversationId: string
-): Promise<boolean> =>
-  (await prisma.conversation.count({
-    where: { id: conversationId, userIds: { has: userId } },
-  })) > 0;
-
-const getAllMessagesByConversationIdWithPagination = async (
+// Fetch messages for the conversation with pagination
+const getMessagesByConversationIdWithPagination = async (
   req: Request,
   res: Response
 ) => {
-  const conversationId = req.query.conversationId as string;
-  const pageSize = parseInt(req.query.pageSize as string, 10) || 10;
-  const cursor = req.query.cursor as string | undefined;
-
-  if (!(await hasConversation(req.auth!.userId, conversationId))) {
-    return res
-      .status(403)
-      .type("text/plain")
-      .send(
-        "Forbidden - User not allowed to retrieve the messages of the conversation."
-      );
-  }
-
-  const allMessages = await prisma.message.findMany({
-    where: {
-      conversation: {
-        id: conversationId,
-      },
-      id: {
-        gt: cursor,
-      },
-    },
-    take: pageSize,
-  });
+  const { userId } = req.auth!; // Get authenticated user's ID
+  const conversationId = req.params.conversationId;
+  const pageSize = parseInt(req.query.pageSize as string, 10) || 10; // Get desired page size from query parameters
+  const cursor = req.query.cursor as string | undefined; // Get optional cursor for pagination from query parameters
 
-  const lastMessage = allMessages[allMessages.length - 1];
-  const nextCursor = lastMessage ? lastMessage.id : null;
+  prisma
+    .$transaction(async (tx) => {
+      try {
+        // Validate conversation existence
+        await tx.conversation.findFirstOrThrow({
+          where: {
+            id: conversationId,
+            userIds: { has: userId },
+          },
+        });
 
-  res.json({
-    allMessages,
-    nextCursor,
-  });
+        const allMessages = await tx.message.findMany({
+          where: {
+            conversation: {
+              id: conversationId, // Filter messages for the conversation
+            },
+            id: {
+              gt: cursor, // Retrieve conversations after the provided cursor (if any)
+            },
+          },
+          orderBy: {
+            createdAt: "asc", // Order messages by creation date in ascending order
+          },
+          take: pageSize, // Limit results to the specified page size
+        });
+
+        // Extract the last messages's cursor for pagination.
+        const lastMessage = allMessages[allMessages.length - 1];
+        const nextCursor = lastMessage ? lastMessage.id : null;
+
+        return res.json({
+          allMessages, // Send the fetched messages in the response
+          nextCursor, // Include the cursor for further pagination
+        });
+      } catch (error) {
+        // Rethrow the error to trigger the transaction rollback
+        throw error;
+      }
```

Comment:
If we're just gonna rethrow without adding any logic then we can remove the try catch all together